test/docker-e2e/e2e_state_sync_test.go

```diff
@@ -214,3 +210,423 @@ func (s *CelestiaTestSuite) TestStateSyncMocha() {
+						break
+					}
+				}
+			}
+			if sequential && heightHistory[len(heightHistory)-1] > 10 {
+				t.Fatalf("EARLY DETECTION: Node appears to be using BLOCK SYNC (sequential height progression: %v)! "+
+					"State sync should jump directly to target height, not increment slowly.", heightHistory)
+			}
+		}
+
+		return !info.CatchingUp && info.LatestBlockHeight >= latestHeight
+	})
+	s.Require().NoError(err, "failed to wait for state sync to complete")
+
+	return heightHistory
+}
+
+// performUpgrade executes the upgrade to the target app version
+func (s *CelestiaTestSuite) performUpgrade(ctx context.Context, chain tastoratypes.Chain, cfg *dockerchain.Config, targetAppVersion uint64) int64 {
+	t := s.T()
+
+	validatorNode := chain.GetNodes()[0]
+	kr := cfg.Genesis.Keyring()
+	records, err := kr.List()
+	s.Require().NoError(err, "failed to list keyring records")
+	s.Require().Len(records, len(chain.GetNodes()), "number of accounts should match number of nodes")
+
+	upgradeHeight := s.signalAndGetUpgradeHeight(ctx, chain, validatorNode, cfg, records, targetAppVersion)
+
+	// Wait for upgrade to complete
+	rpcClient, err := validatorNode.GetRPCClient()
+	s.Require().NoError(err, "failed to get RPC client")
+
+	currentHeight, err := s.GetLatestBlockHeight(ctx, rpcClient)
+	s.Require().NoError(err, "failed to get current height")
+
+	blocksToWait := int(upgradeHeight-currentHeight) + 2
+	t.Logf("Waiting for %d blocks to reach upgrade height %d", blocksToWait, upgradeHeight)
+	s.Require().NoError(wait.ForBlocks(ctx, blocksToWait, chain),
+		"failed to wait for upgrade completion")
+
+	// Verify upgrade completed successfully
+	abciInfo, err := rpcClient.ABCIInfo(ctx)
+	s.Require().NoError(err, "failed to fetch ABCI info")
+	s.Require().Equal(targetAppVersion, abciInfo.Response.GetAppVersion(), "should be at app version %v", targetAppVersion)
+
+	// Produce additional blocks at the target app version (TxSim is still running)
+	t.Logf("Producing 20 more blocks at app version %v", targetAppVersion)
+	s.Require().NoError(wait.ForBlocks(ctx, 20, chain), "failed to wait for post-upgrade blocks")
+
+	return upgradeHeight
+}
+
+// validatorStateSyncAppOverrides modifies the app.toml to configure state sync snapshots for the given node.
+func validatorStateSyncAppOverrides(ctx context.Context, node *tastoradockertypes.ChainNode) error {
+	return config.Modify(ctx, node, "config/app.toml", func(cfg *servercfg.Config) {
+		cfg.StateSync.SnapshotInterval = 5
+		cfg.StateSync.SnapshotKeepRecent = 2
+	})
+}
+
+// validatorStateSyncProducerOverrides configures validators to produce state sync snapshots.
+func validatorStateSyncProducerOverrides(ctx context.Context, node *tastoradockertypes.ChainNode) error {
+	return config.Modify(ctx, node, "config/app.toml", func(cfg *servercfg.Config) {
+		// configure frequent state sync snapshots for new nodes to consume
+		cfg.StateSync.SnapshotInterval = 5
+		cfg.StateSync.SnapshotKeepRecent = 3
+	})
+}
```

Comment:
[question] why are both of these necessary? They seem duplicative.

Reply:
ok will remove it then.